pkg/fs/removeall_windows.go

```diff
@@ -0,0 +1,214 @@
+	if attrs&windows.FILE_ATTRIBUTE_DIRECTORY == 0 {
+		// File
+		if attrs&windows.FILE_ATTRIBUTE_READONLY != 0 {
+			// Read-only flag prevents deletion, so un-set it first.
+			if err := windows.SetFileAttributes(terminate(path), attrs&^windows.FILE_ATTRIBUTE_READONLY); err == windows.ERROR_FILE_NOT_FOUND {
+				return nil
+			} else if err != nil {
+				return pathErr("SetFileAttributes", path, err)
+			}
+		}
+		if err := windows.DeleteFile(terminate(path)); err == windows.ERROR_FILE_NOT_FOUND {
+			return nil
+		} else if err != nil {
+			return pathErr("DeleteFile", path, err)
+		}
+	} else {
+		// Directory
+		// Keep looping, removing children, and attempting to delete again.
+		for {
+			// First, try to delete the directory. This will only work if it's empty.
+			// If that fails then enumerate the entries and delete them first.
+			if err := removeDirectory(terminate(path)); err == nil || err == windows.ERROR_FILE_NOT_FOUND {
+				return nil
+			} else if err != windows.ERROR_DIR_NOT_EMPTY || (attrs&windows.FILE_ATTRIBUTE_REPARSE_POINT != 0 && reparseTag&reparseTagNameSurrogate != 0) {
+				// We either failed for some reason other than the directory not being empty, or because the directory is a name surrogate reparse point.
+				// We don't want to recurse into a name surrogate (e.g. symlink) because we will end up deleting stuff elsewhere on the system.
+				// In this case there's nothing else to do for this entry, so just return an error.
+				return pathErr("RemoveDirectory", path, err)
+			}
+			// Wrapping in an anonymous function so that the deferred FindClose call is scoped properly.
+			// Iterate the directory and remove all children by recursing into removeAll.
+			if err := func() error {
+				var fd findData
+				// findFirstFileExW allows us to avoid the shortname lookup. See comment at top of file for details.
+				pattern := join(path, []uint16{'*'})
+				find, err := findFirstFileExW(terminate(pattern), findExInfoBasic, &fd, findExSearchNameMatch, nil, 0)
+				if err == windows.ERROR_FILE_NOT_FOUND {
+					// No children is weird, because we should always have "." and "..".
+					// If we do hit this, just continue to the next deletion attempt.
+					return nil
+				} else if err != nil {
+					return pathErr("FindFirstFileEx", pattern, err)
+				}
+				defer windows.FindClose(find)
+				for {
+					var child []uint16
+					child, err = truncAtNull(fd.name[:])
+					if err != nil {
+						return err
+					}
+					if !equal(child, []uint16{'.'}) && !equal(child, []uint16{'.', '.'}) { // Ignore "." and ".."
+						if err := removeAll(join(path, child), fd.attributes, fd.reserved0); err != nil {
+							return err
+						}
+					}
+					err = findNextFileW(find, &fd)
+					if err == windows.ERROR_NO_MORE_FILES {
+						break
+					} else if err != nil {
+						return pathErr("FindNextFile", path, err)
+					}
+				}
+				return nil
+			}(); err != nil {
```

Comment:
Can this be moved out to a separate function?
Ideally we can wrap `findFirstFileExW` and `findNextFileW` into a `walkFiles(f func(findData) error)) error`, since that allows the defer to still work, and I am sure we'll need it again in the future

Reply:
I'm not sure here. I think it can be easier to grok the overall flow when it's kept together in a single function, and didn't want to introduce more abstractions like a walk function when the existing code seems fairly straightforward. However the complexity did grow over time, so maybe it would make sense.